api/workers/spotreplace.go

```diff
@@ -0,0 +1,157 @@
+		spotReplace()
+	}
+}
+
+func spotReplace() {
+	log := logger.New("ns=workers.spotreplace").At("spotReplace")
+
+	// do nothing unless autoscaling is on
+	if !spotreplace {
+		return
+	}
+
+	// get system
+	system, err := models.Provider().SystemGet()
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	log.Logf("status=%q", system.Status)
+
+	// only allow running and converging status through
+	switch system.Status {
+	case "running", "converging":
+	default:
+		return
+	}
+
+	resources, err := models.ListResources(os.Getenv("RACK"))
+	if err != nil {
+		return
+	}
+
+	// get on-demand ASG
+	odres, err := models.AutoScaling().DescribeAutoScalingGroups(
+		&autoscaling.DescribeAutoScalingGroupsInput{
+			AutoScalingGroupNames: []*string{
+				aws.String(resources["Instances"].Id),
+			},
+		},
+	)
+	if err != nil {
+		return
+	}
+
+	// count the Healthy, InService on-demand instances
+	onDemandCount := 0
+	for _, onDemandInstance := range odres.AutoScalingGroups[0].Instances {
+		if (*onDemandInstance.HealthStatus == "Healthy") &&
+			((*onDemandInstance.LifecycleState == "InService") || (*onDemandInstance.LifecycleState == "Pending")) {
+			onDemandCount++
+		}
+	}
+
+	onDemandDesiredCapacity := *odres.AutoScalingGroups[0].DesiredCapacity
+
+	// get spot ASG
+	sres, err := models.AutoScaling().DescribeAutoScalingGroups(
+		&autoscaling.DescribeAutoScalingGroupsInput{
+			AutoScalingGroupNames: []*string{
+				aws.String(resources["SpotInstances"].Id),
+			},
+		},
+	)
+	if err != nil {
+		return
+	}
+
+	// count the Healthy, InService spot instances
+	spotCount := 0
+	for _, spotInstance := range sres.AutoScalingGroups[0].Instances {
+		if (*spotInstance.HealthStatus == "Healthy") && (*spotInstance.LifecycleState == "InService") {
+			spotCount++
+		}
+	}
+
+	spotDesiredCapacity := *sres.AutoScalingGroups[0].DesiredCapacity
+	totalInstances := onDemandCount + spotCount
+
+	// if total instances > than InstanceCount and there are too many spot instances, reduse spot count by 1
+	minOnDemandCount := 3
+	if os.Getenv("ON_DEMAND_MIN_COUNT") != "" {
+		minOnDemandCount, err = strconv.Atoi(os.Getenv("ON_DEMAND_MIN_COUNT"))
+		if err != nil {
+			fmt.Println(err)
+		}
+	}
+	if (totalInstances > system.Count) && ((system.Count - spotCount) < minOnDemandCount) {
+		_, err := models.AutoScaling().SetDesiredCapacity(
+			&autoscaling.SetDesiredCapacityInput{
+				AutoScalingGroupName: aws.String(resources["SpotInstances"].Id),
+				DesiredCapacity:      aws.Int64(spotDesiredCapacity - 1),
+			},
+		)
+		if err != nil {
+			fmt.Println(err)
+			return
+		}
+		return
+	}
+
+	// if total instances > than InstanceCount, reduce on-demand desired count by 1
+	if totalInstances > system.Count {
+		newCapacity := int64(onDemandDesiredCapacity - 1)
+		_, err := models.AutoScaling().SetDesiredCapacity(
+			&autoscaling.SetDesiredCapacityInput{
+				AutoScalingGroupName: aws.String(resources["Instances"].Id),
+				DesiredCapacity:      &newCapacity,
```

Comment:
This should be the same aws.Int64 as above, I think there are a couple of others also